### DISCUSSION TITLE
Could Mojo's optimisations (such as kernel fusion) help to optimise general data science workloads?

### DISCUSSION THREAD

**Discussion by JackKelly** [modular/modular#1478] [Q&A] [answered]
If I understand correctly, the main use-case for kernel fusion is for deep learning: Say we have a matrix multiplication followed by a ReLU. Kernel fusion should automatically create a "fused matmul + ReLU" operation.

Could kernel fusion also optimise _general_ data science workloads (which aren't necessarily part of an ML pipeline)? For example, I work with satellite images. I often need to perform a sequence of operations on satellite imagery such as:

1. Select a small crop of the entire dataset.
2. Warp ("geospatially reproject") the image from one map projection to another map projection.
3. Apply a binary mask.
4. Combine multiple channels into one channel (e.g. take the mean pixel value across all channels).
5. Normalise the pixel values so they fall in the range zero to one.

Could kernel fusion "fuse" these sorts of tasks?

More generally: Does Mojo offer performance advantages over C / C++ / Rust for these types of workload? :slightly_smiling_face:

Thanks loads! I'm excited to see how Mojo progresses!

**Comment by ParadaCarleton** [comment]
> More generally: Does Mojo offer performance advantages over C / C++ / Rust for these types of workload? 🙂

When I asked a similar question on the Discord about using Mojo for data science (and asked about some features I'd like for similar workflows), the impression I got is they're not interested in data science or scientific/numerical computing. If that's correct, I think PyTorch, Julia, or JAX would be a better fit.

However, you're 100% correct that kernel fusion is very important to performance in general data science workloads (and is implemented as part of Julia's broadcasting system). There's not really a computer-level distinction between numerical computing and ML--both tend to need the same kinds of optimizations.

**Comment by iamtimdavis** [comment, marked as answer]
It's definitely not correct to say "we aren't interested in data science or scientific/numerical computing" - of course we are, its a critical part of AI. So to make it very clear - yes, we are deeply interested in data science and numerical computing and would love to see Mojo grow here over time.

I again implore you, please don't make this a "Mojo vs Julia" discussion - we've answered this question many times and you can see our thoughts here - https://docs.modular.com/mojo/faq.html#why-not-make-julia-better

To answer @JackKelly original question - we'd encourage you to play with the MAX Engine (https://www.modular.com/max) when the Developer Edition arrives in Q1. You can load custom operations directly into the engine for fusion opportunities.

**Reply by ParadaCarleton** [reply]
> I again implore you, please don't make this a "Mojo vs Julia" discussion - we've answered this question many times and you can see our thoughts here - https://docs.modular.com/mojo/faq.html#why-not-make-julia-better

Just wanted to note I wasn't trying to make this a "which is better" discussion, just that this was the impression I got from conversations in the Discord--that Mojo is something very different from either Julia or Python (like the FAQ you linked says), and it's supposed to be a backend/systems programming language like C++ instead of a frontend language for data science.

If that's not the case, I'm definitely surprised, because the Mojo team doesn't seem to have many (any?) data scientists or scientists on it, and the language seems to be very focused on memory safety, low-level bit fiddling, traits, and keeping the language familiar to C++ devs. It's also the impression of most of the data scientists I've talked to about this, mostly from Chris Lattner's comments on Mojo replacing C++, not Python.
